### PR TITLE
Generate documentation of ServiceDiscovery automatically

### PR DESCRIPTION
**Pull request by 315157973**
Master Issue: #10041

### Motivation
The config object in the Service Discovery module has been annotated.
So the document can be automatically generated.
But this module and Pulsar proxy module both depend on broker-common, so we need to move the common CMD to the common module, otherwise we need to copy the code.

### Modifications
1) Move common CMD to broker-common
2) Add unit test for  Service Discovery

### CODE REVIEW

**Comment by 315157973** [comment]
# Service discovery
|Name|Description|Default|
|---|---|---|
| zookeeperServers | Local-Zookeeper quorum connection string | null | 
| configurationStoreServers | Configuration Store connection string | null | 
| zookeeperSessionTimeoutMs | ZooKeeper session timeout in ms | 30000 | 
| zooKeeperCacheExpirySeconds | ZooKeeper cache expiry time in seconds | 300 | 
| servicePort | Port to use to server binary-proto request | Optional[5000] | 
| servicePortTls | Port to use to server binary-proto-tls request | Optional.empty | 
| webServicePort | Port to use to server HTTP request | Optional[8080] | 
| webServicePortTls | Port to use to server HTTPS request | Optional.empty | 
| bindOnLocalhost | Control whether to bind directly on localhost rather than on normal hostname | false | 
| superUserRoles | Role names that are treated as "super-user", meaning they will be able to do all admin operations and publish/consume from all topics | [] | 
| authorizationAllowWildcardsMatching | Allow wildcard matching in authorization (wildcard matching only applicable if wildcard-char: * presents at first or last position eg: *.pulsar.service, pulsar.service.* | false | 
| authenticationEnabled | Whether enable authentication | false | 
| authenticationProviders | Authentication provider name list, which is a list of class names | [] | 
| authorizationEnabled | Whether Enforce authorization | false | 
| authorizationProvider | Authorization provider fully qualified class-name | org.apache.pulsar.broker.authorization.PulsarAuthorizationProvider | 
| tlsCertRefreshCheckDurationSec | Tls cert refresh duration in seconds (set 0 to check on every new connection) | 300 | 
| tlsCertificateFilePath | Path for the TLS certificate file | null | 
| tlsKeyFilePath | Path for the TLS private key file | null | 
| tlsTrustCertsFilePath | Path for the trusted TLS certificate file |  | 
| tlsAllowInsecureConnection | Accept untrusted TLS certificate from client | false | 
| tlsProtocols | Specify the tls protocols the broker will use to negotiate during TLS Handshake. Example:- [TLSv1.3, TLSv1.2] | [] | 
| tlsCiphers | Specify the tls cipher the broker will use to negotiate during TLS Handshake. Example:- [TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256] | [] | 
| tlsRequireTrustedClientCertOnConnect | Specify whether Client certificates are required for TLS. Reject the Connection if the Client Certificate is not trusted. | false | 
| tlsEnabledWithKeyStore | Enable TLS with KeyStore type configuration in broker | false | 
| tlsProvider | Full class name of TLS Provider | null | 
| tlsKeyStoreType | TLS KeyStore type configuration in broker: JKS, PKCS12 | JKS | 
| tlsKeyStore | TLS KeyStore path in broker | null | 
| tlsKeyStorePassword | TLS KeyStore password in broker | null | 
| tlsTrustStoreType | TLS TrustStore type configuration in broker: JKS, PKCS12 | JKS | 
| tlsTrustStore | TLS TrustStore path in broker | null | 
| tlsTrustStorePassword | TLS TrustStore password in broker | null | 
| properties | Can store string in key-value format | {} |

File: pulsar-discovery-service/src/main/java/org/apache/pulsar/discovery/service/server/ServiceConfig.java [hunk 1]
```diff
@@ -36,91 +37,209 @@
...
+    @ApiModelProperty(
+            name = "zooKeeperCacheExpirySeconds",
+            value = "ZooKeeper cache expiry time in seconds"
+    )
     private int zooKeeperCacheExpirySeconds=300;
 
-    // Port to use to server binary-proto request
+    @ApiModelProperty(
+            name = "servicePort",
+            value = "Port to use to server binary-proto request"
```
Comment (Anonymitaet) [inline pulsar-discovery-service/src/main/java/org/apache/pulsar/discovery/service/server/ServiceConfig.java:69]:
```suggestion
            value = "Port used to server binary-proto request"
```

Comment (Anonymitaet) [reply pulsar-discovery-service/src/main/java/org/apache/pulsar/discovery/service/server/ServiceConfig.java:69]: Please update all occurrences.

File: pulsar-discovery-service/src/main/java/org/apache/pulsar/discovery/service/server/ServiceConfig.java [hunk 2]
```diff
@@ -36,91 +37,209 @@
...
+    @ApiModelProperty(
+            name = "webServicePort",
+            value = "Port to use to server HTTP request"
+    )
     private Optional<Integer> webServicePort = Optional.ofNullable(8080);
-    // Port to use to server HTTPS request
+
+    @ApiModelProperty(
+            name = "webServicePortTls",
+            value = "Port to use to server HTTPS request"
```
Comment (Anonymitaet) [inline pulsar-discovery-service/src/main/java/org/apache/pulsar/discovery/service/server/ServiceConfig.java:87]: Please update all occurrences.

File: pulsar-discovery-service/src/main/java/org/apache/pulsar/discovery/service/server/ServiceConfig.java [hunk 3]
```diff
@@ -36,91 +37,209 @@
...
+            value = "Control whether to bind directly on localhost rather than on normal hostname"
+    )
     private boolean bindOnLocalhost = false;
 
-    // Role names that are treated as "super-user", meaning they will be able to
-    // do all admin operations and publish/consume from all topics
+    @ApiModelProperty(
+            name = "superUserRoles",
+            value = "Role names that are treated as \"super-user\", meaning they will be able to "
+                    + "do all admin operations and publish/consume from all topics"
```
Comment (Anonymitaet) [inline pulsar-discovery-service/src/main/java/org/apache/pulsar/discovery/service/server/ServiceConfig.java:100]:
```suggestion
            value = "Role names that are treated as \"super-user\", meaning they are able to "
                    + "do all admin operations and publish to or consume from all topics"
```

File: pulsar-discovery-service/src/main/java/org/apache/pulsar/discovery/service/server/ServiceConfig.java [hunk 4]
```diff
@@ -36,91 +37,209 @@
...
+    )
     private Set<String> superUserRoles = Sets.newTreeSet();
 
-    // Allow wildcard matching in authorization
-    // (wildcard matching only applicable if wildcard-char:
-    // * presents at first or last position eg: *.pulsar.service, pulsar.service.*)
+    @ApiModelProperty(
+            name = "authorizationAllowWildcardsMatching",
+            value = "Allow wildcard matching in authorization (wildcard matching only applicable "
+                    + "if wildcard-char: * presents at first or last position eg: *.pulsar.service, pulsar.service.*"
```
Comment (Anonymitaet) [inline pulsar-discovery-service/src/main/java/org/apache/pulsar/discovery/service/server/ServiceConfig.java:107]:
```suggestion
            value = "Allow wildcard matching in authorization (wildcard matching only applicable "
                    + "if wildcard char * presents at first or last position. For example, *.pulsar.service, pulsar.service.*"
```

File: pulsar-discovery-service/src/main/java/org/apache/pulsar/discovery/service/server/ServiceConfig.java [hunk 5]
```diff
@@ -36,91 +37,209 @@
...
+    @ApiModelProperty(
+            name = "authenticationProviders",
+            value = "Authentication provider name list, which is a list of class names"
+    )
     private Set<String> authenticationProviders = Sets.newTreeSet();
-    // Enforce authorization
+
+    @ApiModelProperty(
+            name = "authorizationEnabled",
+            value = "Whether Enforce authorization"
```
Comment (Anonymitaet) [inline pulsar-discovery-service/src/main/java/org/apache/pulsar/discovery/service/server/ServiceConfig.java:125]:
```suggestion
            value = "Whether enforce authorization"
```

File: pulsar-discovery-service/src/main/java/org/apache/pulsar/discovery/service/server/ServiceConfig.java [hunk 6]
```diff
@@ -36,91 +37,209 @@
...
     private String authorizationProvider = PulsarAuthorizationProvider.class.getName();
 
     /***** --- TLS --- ****/
     @Deprecated
     private boolean tlsEnabled = false;
-    // Tls cert refresh duration in seconds (set 0 to check on every new connection)
+
+    @ApiModelProperty(
+            name = "tlsCertRefreshCheckDurationSec",
+            value = "Tls cert refresh duration in seconds (set 0 to check on every new connection)"
```
Comment (Anonymitaet) [inline pulsar-discovery-service/src/main/java/org/apache/pulsar/discovery/service/server/ServiceConfig.java:141]:
```suggestion
            value = "TLS cert refresh duration (in seconds). 0 means checking every new connection."
```

File: pulsar-discovery-service/src/main/java/org/apache/pulsar/discovery/service/server/ServiceConfig.java [hunk 7]
```diff
@@ -36,91 +37,209 @@
...
+            value = "Accept untrusted TLS certificate from client"
+    )
     private boolean tlsAllowInsecureConnection = false;
-    // Specify the tls protocols the broker will use to negotiate during TLS Handshake.
-    // Example:- [TLSv1.3, TLSv1.2]
+
+    @ApiModelProperty(
+            name = "tlsProtocols",
+            value = "Specify the tls protocols the broker will use to negotiate during TLS Handshake. "
+                    + "Example:- [TLSv1.3, TLSv1.2]"
```
Comment (Anonymitaet) [inline pulsar-discovery-service/src/main/java/org/apache/pulsar/discovery/service/server/ServiceConfig.java:172]:
```suggestion
            value = "Specify the TLS protocols the broker uses to negotiate during TLS Handshake. "
                    + "Example:- [TLSv1.3, TLSv1.2]"
```

Comment (Anonymitaet) [reply pulsar-discovery-service/src/main/java/org/apache/pulsar/discovery/service/server/ServiceConfig.java:172]: What's the meaning of `-` after `Example:`?

File: pulsar-discovery-service/src/main/java/org/apache/pulsar/discovery/service/server/ServiceConfig.java [hunk 8]
```diff
@@ -36,91 +37,209 @@
...
+    )
     private Set<String> tlsProtocols = Sets.newTreeSet();
-    // Specify the tls cipher the broker will use to negotiate during TLS Handshake.
-    // Example:- [TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256]
+
+    @ApiModelProperty(
+            name = "tlsCiphers",
+            value = "Specify the tls cipher the broker will use to negotiate during TLS Handshake. "
+                    + "Example:- [TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256]"
```
Comment (Anonymitaet) [inline pulsar-discovery-service/src/main/java/org/apache/pulsar/discovery/service/server/ServiceConfig.java:179]: same

File: pulsar-discovery-service/src/main/java/org/apache/pulsar/discovery/service/server/ServiceConfig.java [hunk 9]
```diff
@@ -36,91 +37,209 @@
...
+    )
     private Set<String> tlsCiphers = Sets.newTreeSet();
-    // Specify whether Client certificates are required for TLS
-    // Reject the Connection if the Client Certificate is not trusted.
+
+    @ApiModelProperty(
+            name = "tlsRequireTrustedClientCertOnConnect",
+            value = "Specify whether Client certificates are required for TLS. "
+                    + "Reject the Connection if the Client Certificate is not trusted."
```
Comment (Anonymitaet) [inline pulsar-discovery-service/src/main/java/org/apache/pulsar/discovery/service/server/ServiceConfig.java:186]:
```suggestion
            value = "Specify whether client certificates are required for TLS. "
                    + "Reject the connection if the client certificate is not trusted."
```

File: pulsar-discovery-service/src/main/java/org/apache/pulsar/discovery/service/server/ServiceConfig.java [hunk 10]
```diff
@@ -36,91 +37,209 @@
...
+    @ApiModelProperty(
+            name = "tlsProvider",
+            value = "Full class name of TLS Provider"
+    )
     private String tlsProvider = null;
-    // TLS KeyStore type configuration in broker: JKS, PKCS12
+
+    @ApiModelProperty(
+            name = "tlsKeyStoreType",
+            value = "TLS KeyStore type configuration in broker: JKS, PKCS12"
```
Comment (Anonymitaet) [inline pulsar-discovery-service/src/main/java/org/apache/pulsar/discovery/service/server/ServiceConfig.java:205]:
```suggestion
            value = "TLS KeyStore type configurations in broker are JKS or PKCS12"
```
do you mean this?

File: pulsar-discovery-service/src/main/java/org/apache/pulsar/discovery/service/server/ServiceConfig.java [hunk 11]
```diff
@@ -36,91 +37,209 @@
...
+    @ApiModelProperty(
+            name = "tlsKeyStorePassword",
+            value = "TLS KeyStore password in broker"
+    )
     private String tlsKeyStorePassword = null;
-    // TLS TrustStore type configuration in broker: JKS, PKCS12
+
+    @ApiModelProperty(
+            name = "tlsTrustStoreType",
+            value = "TLS TrustStore type configuration in broker: JKS, PKCS12"
```
Comment (Anonymitaet) [inline pulsar-discovery-service/src/main/java/org/apache/pulsar/discovery/service/server/ServiceConfig.java:223]: same

File: pulsar-discovery-service/src/main/java/org/apache/pulsar/discovery/service/server/ServiceConfig.java [hunk 12]
```diff
@@ -36,91 +37,209 @@
...
+
+    @ApiModelProperty(
+            name = "tlsTrustStorePassword",
+            value = "TLS TrustStore password in broker"
+    )
     private String tlsTrustStorePassword = null;
 
+    @ApiModelProperty(
+            name = "properties",
+            value = "Can store string in key-value format"
```
Comment (Anonymitaet) [inline pulsar-discovery-service/src/main/java/org/apache/pulsar/discovery/service/server/ServiceConfig.java:241]:
```suggestion
            value = "You can store string in key-value format"
```

**Review by Anonymitaet — CHANGES_REQUESTED** [review]

**Review by Anonymitaet — APPROVED** [review]
Approve from the language perspective.

**Review by tuteng — APPROVED** [review]
LGTM

**Review by sijie — APPROVED** [review]

**Pull request merged**